### PR TITLE
Image server side

### PR DESCRIPTION
**Pull request by gtalarico**
# Improve Server-Side Image Handling

This PR was started based on this [Notion Doc](https://www.notion.so/gtalarico/Resize-Image-on-Server-7d2cbe4d71494b67a7426f42a28c51f1). 

I am also settings a terrible example here with major scope creep in this PR 🤦

* Adds the ability to resize images on upload endpoint
* Improves how we reference images in models (`ImageAsset` obj instead of string url)
* Adds cover and images to Post model

Previously, Company logos and cover images had a weird reference to images:
* When creating a new company for example, (`/images/upload/`) would create a new `Image` and return the url (s3 url)
* URLs were then saved as `cover_url`, and `logo_url` fields. This made it hard to know created "Image" assets were linked or even used or to store multiple versions and sizes of those images.

The changes here proposed here use a new renamed model, "ImageAsset", which is referenced as a foreign key instead. 
This model will also be used to store any type of image content upload by users.

The new Image model also allow us to add hooks to auto-generate thumbnail or other sizes using signals or resize function. This PR does not add any auto-sizing, only the capability to.

### In Progress
Before updating how images models are related throughout the api, I am testing the new data model with the `Post` model only, adding a new field `cover_img` and `images`. If we are happy with this I will likely apply this to `Company` and `Profile` image fields as well.

### Note on Post Images
When posts are created, images are base64 encoded as data-uris by quill js editor. This PR also adds a service that extracts those images from image data-uri src, creates `ImageAssets`, and swaps `src` from body to use s3 url.

### TODO

- [ ] Replace `Company.cover_url: str` and `Company.logo_url` `ImageAsset`
- [ ] Create Frontend PR to update company assets upload and use `?width=` and `height=` params to resize images
- [ ] Create script to batch Resize some super large images already in Prod?

### Post Create Image Extraction Example
Post below shows how a post with embedded images automatically gets a cover image and an image asset object created.
image

----
# Image Sizes
We have not defined a consistent standard for image sizing yet. Below are some examples

| Place | Current Size|  Future Size | 
|-------|------|--------------|
| Company Card | 512x200 (2.56:1)  | ? |
| Company Card Cropper |  600x300 (2:1) | ? |
| Company Logo| 80x80 (1:1)| ? |
| Profile List| 30x30 (1:1) | ? |
| Profile Nav| 24x24 (1:1) | ? |
| Profile Page | 100x100 (1:1) | ? |
| Planned People List |  58x58 (1:1)  |  ? |

#### Company Card
image

#### Company Card Crop
image

#### Company Logo
image

#### Profile Avatar 
image

#### Profile Login 
image

#### Peopl Page
image

#### Planned People List Page
image

### CODE REVIEW

File: api/aecworks/settings.py [hunk 1]
```diff
@@ -69,6 +69,7 @@
     "django_extensions",
     "debug_toolbar",
     "django_celery_results",
+    "django_cleanup",
```
Comment (gtalarico) [inline api/aecworks/settings.py:72]: ext. auto deletes images from bucket when the corresponding model is deleted
https://github.com/un1t/django-cleanup

**Comment by danrocha** [comment]
This is great. So the goal would be to use the width and height url param to request images of any size?
Or there would be pre-defined image sizes which we can request?

**Comment by gtalarico** [comment]
> This is great. So the goal would be to use the width and height url param to request images of any size?
Or there would be pre-defined image sizes which we can request?

This would be to let us resize image on write, so for example company logo post would resize to 80x80, 
cover to 600x300, etc.
We can write the largest needed size.

Once we have that, we can also easily add https://github.com/respondcreate/django-versatileimagefield
to support request images of any size. I will do this next

**Comment by danrocha** [comment]
> We can write the largest needed size.

Great. We should then try to consider retina displays and write everything at least 2x the maximum size we want ;)

**Comment by gtalarico** [comment]
> Great. We should then try to consider retina displays and write everything at least 2x the maximum size we want ;)


yes! So "future size" below should be:
1. more consolidated, less variations
2. big enough for retina. How big is big enough @danrocha ? 
eg. company card is 80-90% width on mobile, but desktop layout maxes out at 900px.
 

| Place | Current Size|  Future Size | 
|-------|------|--------------|
| Company Card | 512x200 (2.56:1)  | ? |
| Company Card Cropper |  600x300 (2:1) | ? |
| Company Logo| 80x80 (1:1)| ? |
| Profile List| 30x30 (1:1) | ? |
| Profile Nav| 24x24 (1:1) | ? |
| Profile Page | 100x100 (1:1) | ? |
| Planned People List |  58x58 (1:1)  |  ? |

**Comment by danrocha** [comment]
Ok, so most retina displays are DPR 2, which means images should be 2x larger than in normal displays. We could also push for DPR 3 in case possible (3x).

Why don't we simplify as this (for max sizes):

Place | Max Size | Additional sizes
-- | -- | --
Company Card & Cropper | 1200x600 (DPR2) | 600x300 (DPR1)
Company Logo | 160x160 (DPR2) | 80x80 (DPR1)
Profile* |  200x200 | 116x116 (list new DPR2) 100x100 (profile page DPR1) 58x58 (list new DPR1) 48x48 (nav DPR2) 24x24 (nav DPR1)

* but do we have control over this? isn't it coming from oauth?

**Pull request merged**